src/caffe/test/test_roi_pooling_layer.cpp

```diff
@@ -0,0 +1,94 @@
+#include <cstdlib>
+#include <cstring>
+#include <vector>
+
+#include "boost/scoped_ptr.hpp"
+#include "gtest/gtest.h"
+
+#include "caffe/blob.hpp"
+#include "caffe/common.hpp"
+#include "caffe/filler.hpp"
+#include "caffe/layers/roi_pooling_layer.hpp"
+#include "caffe/test/test_caffe_main.hpp"
+#include "caffe/test/test_gradient_check_util.hpp"
+
+using boost::scoped_ptr;
+
+namespace caffe {
+
+#ifndef CPU_ONLY
+template <typename Dtype>
+class ROIPoolingLayerTest : public GPUDeviceTest<Dtype> {
+ protected:
+  ROIPoolingLayerTest()
+      : blob_bottom_data_(new Blob<Dtype>(4, 3, 12, 8)),
+        blob_bottom_rois_(new Blob<Dtype>(4, 5, 1, 1)),
+        blob_top_data_(new Blob<Dtype>()) {
+    // fill the values
+    FillerParameter filler_param;
+    filler_param.set_std(10);
+    GaussianFiller<Dtype> filler(filler_param);
+    filler.Fill(this->blob_bottom_data_);
+
+    blob_bottom_vec_.push_back(blob_bottom_data_);
+
+    int i = 0;
+    // caffe_rng_rand() % 4;
+    blob_bottom_rois_->mutable_cpu_data()[0 + 5*i] = 0;
+    blob_bottom_rois_->mutable_cpu_data()[1 + 5*i] = 1;  // x1 < 8
+    blob_bottom_rois_->mutable_cpu_data()[2 + 5*i] = 1;  // y1 < 12
+    blob_bottom_rois_->mutable_cpu_data()[3 + 5*i] = 6;  // x2 < 8
+    blob_bottom_rois_->mutable_cpu_data()[4 + 5*i] = 6;  // y2 < 12
+    i = 1;
+    blob_bottom_rois_->mutable_cpu_data()[0 + 5*i] = 2;
+    blob_bottom_rois_->mutable_cpu_data()[1 + 5*i] = 6;  // x1 < 8
+    blob_bottom_rois_->mutable_cpu_data()[2 + 5*i] = 2;  // y1 < 12
+    blob_bottom_rois_->mutable_cpu_data()[3 + 5*i] = 7;  // x2 < 8
+    blob_bottom_rois_->mutable_cpu_data()[4 + 5*i] = 11;  // y2 < 12
+    i = 2;
+    blob_bottom_rois_->mutable_cpu_data()[0 + 5*i] = 1;
+    blob_bottom_rois_->mutable_cpu_data()[1 + 5*i] = 3;  // x1 < 8
+    blob_bottom_rois_->mutable_cpu_data()[2 + 5*i] = 1;  // y1 < 12
+    blob_bottom_rois_->mutable_cpu_data()[3 + 5*i] = 5;  // x2 < 8
+    blob_bottom_rois_->mutable_cpu_data()[4 + 5*i] = 10;  // y2 < 12
+    i = 3;
+    blob_bottom_rois_->mutable_cpu_data()[0 + 5*i] = 0;
+    blob_bottom_rois_->mutable_cpu_data()[1 + 5*i] = 3;  // x1 < 8
+    blob_bottom_rois_->mutable_cpu_data()[2 + 5*i] = 3;  // y1 < 12
+    blob_bottom_rois_->mutable_cpu_data()[3 + 5*i] = 3;  // x2 < 8
+    blob_bottom_rois_->mutable_cpu_data()[4 + 5*i] = 3;  // y2 < 12
+
+    blob_bottom_vec_.push_back(blob_bottom_rois_);
+    blob_top_vec_.push_back(blob_top_data_);
+  }
+  virtual ~ROIPoolingLayerTest() {
+    delete blob_bottom_data_;
+    delete blob_bottom_rois_;
+    delete blob_top_data_;
+  }
+  Blob<Dtype>* const blob_bottom_data_;
+  Blob<Dtype>* const blob_bottom_rois_;
+  Blob<Dtype>* const blob_top_data_;
+  vector<Blob<Dtype>*> blob_bottom_vec_;
+  vector<Blob<Dtype>*> blob_top_vec_;
+};
+
+// Forward and Backward CPU not implemented yet
+TYPED_TEST_CASE(ROIPoolingLayerTest, TestDtypes);
+
+TYPED_TEST(ROIPoolingLayerTest, TestGradient) {
+  typedef typename TypeParam::Dtype Dtype;
```

Comment:
This line is where the compiler is complaining. TypeParam here is the actually data type, since you specified GPUDeviceTest. Remove this line. And change <Dtype> to  <TypeParam> in line 87 and 88.